be/src/exec/vectorized/olap_scan_node.cpp

```diff
@@ -479,8 +485,19 @@ Status OlapScanNode::_start_scan_thread(RuntimeState* state) {
 
     _dict_optimize_parser.rewrite_conjuncts<true>(&conjunct_ctxs, state);
 
-    int scanners_per_tablet = std::max(1, 64 / (int)_scan_ranges.size());
-    for (auto& scan_range : _scan_ranges) {
+    int tablet_count = _scan_ranges.size();
+    for (int k = 0; k < tablet_count; ++k) {
+        auto& scan_range = _scan_ranges[k];
+        auto& tablet_rowset = _tablet_rowsets[k];
+
+        size_t segment_nums = 0;
+        for (const auto& rowset : tablet_rowset) {
+            segment_nums += rowset->num_segments();
+        }
+        COUNTER_UPDATE(_segment_counter, segment_nums);
+
+        int scanners_per_tablet = std::min(segment_nums, kMaxScannerPerRange / _scan_ranges.size());
+
         int num_ranges = key_ranges.size();
         int ranges_per_scanner = std::max(1, num_ranges / scanners_per_tablet);
```

Comment:
if `num_ranges` is not divisible by `scanners_per_tablet`, the number of scanners may be also greater than 64.
For example, assume that `num_tablets = 1` `scanners_per_tablet=64` and `num_ranges=70`, `ranges_per_scanner` will be 1 and there will be 70 scanners.

Maybe we should assign 2 ranges to the first 6 scanner, and 1 range for the other scanners?

Reply:
each scanner could only handle TabletReaderParams, so each scanner could only 'le' or 'lt' range. so it's hard to split range to scanner. later I'll try to refactor this.